### PR TITLE
Proposal for TCP server at all layers (core, transport & rxjava1)

### PR DESCRIPTION
**Pull request by NiteshKant**
Based on various discussions around different issues, I am proposing a fresh design for a TCP server in this PR. 
I have implemented a TCP server at all layers in reactive-ipc, i.e. core, transport(netty) and rxjava1. I have also removed all code which isn't related to the implementation. We can discuss those utilities in isolation as and when required to be added. Although, not used, I have kept the `Buffer` and related classes as that aspect is not yet discussed to conclusion.

There are two samples each for netty transport and rxjava1 layer available to see how will a typical interaction look like in modules ripc-transport-netty4-examples and ripc-rxjava1-examples

Backpressure has not been implemented yet.

### CODE REVIEW

**Comment by benjchristensen** [comment]
@NiteshKant Can you rebase this for consideration to merge?

**Comment by benjchristensen** [comment]
@jbrisbin @smaldini This feels like a good first step towards getting a protocol working end-to-end with all the layers (transport, core, adaptor) and example code.

There is a lot to discuss and this is far from production worth, and the APIs are most definitely not all correct, but I propose we merge this and use it to start iterating with further pull requests and discussion.

**Comment by benjchristensen** [comment]
BTW, once we have a good approach we can migrate the tcp protocol code to its own project as per https://github.com/reactive-ipc/reactive-ipc-jvm/issues/18

**Comment by NiteshKant** [comment]
@benjchristensen the code seems to be on the latest from upstream and is ready to be merged.

**Comment by jbrisbin** [comment]
:+1:

**Comment by benjchristensen** [comment]
@NiteshKant The question about rebasing was whether we should retain the 7 commits for this, or squash to 1 commit.

**Comment by jbrisbin** [comment]
Ideally they should be squashed to 1.

**Comment by NiteshKant** [comment]
Looks like my attempt to squash across merge commits rendered this PR unusable :(
Lemme try to fix it

**Comment by benjchristensen** [comment]
You shouldn't need the merge commit. You can just rebase on top of the master branch. Or just checkout master again from scratch, branch, then copy/paste your code and commit a single clean commit.

File: ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpHandler.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package io.ripc.protocol.tcp;
+
+import org.reactivestreams.Publisher;
+
+public interface TcpHandler<R, W> {
```
Comment (jbrisbin) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpHandler.java:5]: As a general rule, I'm not a fan of duplicating names in both packages and class names. It also doesn't "handle" `Tcp`, it handles a `Connection`. If needing a qualifier to `Handler`, it should probably be `ConnectionHandler`.

File: ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package io.ripc.protocol.tcp;
+
+public interface TcpInterceptor<I, O, II, OO> {
```
Comment (jbrisbin) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpInterceptor.java:3]: This shouldn't have 4 generics placeholders. 2 is sufficient with 2 transformative ones on the method. Using this in practice will be horrific if the names are of any length.

File: build.gradle [hunk 3]
```diff
@@ -98,17 +113,32 @@ project('ripc-transport-netty4') {
 
 		// Netty
 		compile "io.netty:netty-all:$nettyVersion"
+        compile "org.slf4j:slf4j-api:1.7.6"
 	}
 }
 
-project('ripc-composition-rxjava1') {
+project('ripc-rxjava1') {
```
Comment (jbrisbin) [inline build.gradle:119]: Taking out the reference to "composition" doesn't tell me what this module's responsibility is now.

File: build.gradle [hunk 4]
```diff
@@ -87,9 +82,29 @@ project('ripc-protocol-tcp') {
 	dependencies {
 		// ripc-core
 		compile project(":ripc-core")
+        compile "org.slf4j:slf4j-api:1.7.6"
+		testCompile 'io.reactivex:rxjava:1.0.8'
```
Comment (jbrisbin) [inline build.gradle]: IMO We shouldn't mix layers. We should provide mocks and stick to RS APIs in layers other than composition.

**Comment by jbrisbin** [comment]
FWIW @rstoyanchev @smaldini and I are going to be going over this in a more comprehensive way today as we individually evaluate some of these approaches and then get together to discuss them to offer more comprehensive feedback than is possible in line comments.

File: ripc-core/src/main/java/io/ripc/internal/Publishers.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+public class Publishers {
+
+    public static <T> Publisher<T> just(final T value) {
+        return new Publisher<T>() {
+            @Override
+            public void subscribe(final Subscriber<? super T> s) {
+                s.onSubscribe(new Subscription() {
+                    @Override
+                    public void request(long n) {
+                        s.onNext(value);
```
Comment (jbrisbin) [inline ripc-core/src/main/java/io/ripc/internal/Publishers.java:19]: There should be a positive demand check here per the spec. The `Specifications` class was intended to be the helper that aggregated these checks. This will be an ongoing issue (proper adherence to the spec) any time we implement helpers in the core.

File: ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpHandler.java [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+
+    Publisher<Void> handle(Connection<R, W> connection);
```
Comment (jbrisbin) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpHandler.java]: What does the returned `Publisher<Void>` signal to a `Subscriber` via `onComplete`? What happens if multiple handlers are added? Does each `Subscriber` receive an `onComplete` signal when the peer terminates?

File: ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpInterceptor.java [hunk 7]
```diff
@@ -0,0 +1,7 @@
...
+
+    TcpHandler<II, OO> intercept(TcpHandler<I, O> handler);
```
Comment (jbrisbin) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpInterceptor.java:5]: I was envisioning interceptors being analogous to Netty's pipeline feature. Pipelines can be dynamically altered and new ones are created when new connections are made. How would we use this intercept mechanism to wrap each new connection dynamically (especially those that come in as a result of an automatic reconnect after a client connection has died)?

File: ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpServer.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        if (null == thisHandler) {
+            doStart(handler);
+        } else {
+            lazyHandler.start(handler);
+            doStart(thisHandler);
+        }
+        return this;
+    }
+
+    public final void startAndAwait(TcpHandler<R, W> handler) {
```
Comment (jbrisbin) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpServer.java:24]: Should we be mixing blocking semantics into a fundamentally non-blocking server? It's likely that few "real" situations would use this anyway since there would be other means of keeping the JVM alive (a la Spring Boot, which would definitely not use an await() method. Can't we accomplish this via helpers that take a `Publisher` and block until `Subscriber.onComplete`?

File: ripc-rxjava1/src/main/java/io/ripc/rx/protocol/tcp/RxConnection.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    /**
+     * Writes the passed stream of {@code data} and returns the result as an {@link Observable}. All items emitted by
+     * this stream are flushed on completion of the stream.
+     *
+     * @param data Data stream to write.
+     *
+     * @return Result of write.
+     */
+    public Observable<Void> write(Observable<W> data) {
+        return toObservable(delegate.write(toPublisher(data)));
```
Comment (jbrisbin) [inline ripc-rxjava1/src/main/java/io/ripc/rx/protocol/tcp/RxConnection.java:39]: Does this mean every write results in 2 allocations in addition to the data `Observable`? I'm not familiar with how Rx does this back-and-forth between plain Rx and Reactive Streams.

**Comment by jbrisbin** [comment]
How would we handle connection close events and reconnects?

**Comment by NiteshKant** [comment]
> > FWIW @rstoyanchev @smaldini and I are going to be going over this in a more comprehensive way today as we individually evaluate some of these approaches and then get together to discuss them to offer more comprehensive feedback than is possible in line comments.

All great questions @jbrisbin . I will reply to them so that you guys have more context on these when you discuss them.

File: ripc-rxjava1/src/main/java/io/ripc/rx/protocol/tcp/RxConnection.java (hunk 9 above)
Comment (NiteshKant) [inline ripc-rxjava1/src/main/java/io/ripc/rx/protocol/tcp/RxConnection.java:39]: Since they (`Publisher` and `Observable`) are different objects with no common interface, I would imagine an allocation per transformation between them is always required. Do you have a better alternative?

File: ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpServer.java (hunk 8 above)
Comment (NiteshKant) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpServer.java:24]: Me & @benjchristensen talked about it and unless we model the server as a `Publisher` of `ServerState` which is `Start` and `Stop` it seems like an overkill since that would mean someone has to subscribe to the state observable to start the server. Since, there will only be two state `Start` and `Stop`, it sounds unnecessary to me to model it that way.
Are there any other usecases apart from await that you would see us gaining from if we move to returning a `Publisher` from `start()`?

File: ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpInterceptor.java (hunk 7 above)
Comment (NiteshKant) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpInterceptor.java:5]: > > I was envisioning interceptors being analogous to Netty's pipeline feature.

Yes and No. Yes because it can change the types of input and output on the channel (read and write in netty pipeline) and no because it does not intercept any other channel lifecycle event apart from read/write.
Conceptually, interceptors are application level constructs where as netty channels are network level constructs thus needing more feedback around channel lifecycle.

> > How would we use this intercept mechanism to wrap each new connection dynamically

Have a look at the example [here](https://github.com/reactive-ipc/reactive-ipc-jvm/pull/19/files#diff-6327a044d2d50969539215f133a15744R19) which is wrapping over the next interceptor/handler in the chain and hence is in complete control of what connection to pass downstream.

> > especially those that come in as a result of an automatic reconnect after a client connection has died

I did not quiet understand, can you elaborate?

File: ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpHandler.java (hunk 6 above)
Comment (NiteshKant) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpHandler.java]: This is a server construct. One server can only ever have a single handler.

> Does each Subscriber receive an onComplete signal when the peer terminates?

The only subscriber to this `Publisher<Void>` is reactive-ipc internals. No connection lifecycle events are piped to this, unless the handle implementation choose to do so. eg:

This does not wire the connection close.

``` java
        Netty4TcpServer.create(0)
                       .start(connection -> Publishers.error(new IllegalStateException()));
```

OTOH, this does as it is wiring the write to the returned `Publisher` which fails when connection is closed:

``` java
        Netty4TcpServer.create(0)
                       .start(connection -> connection.write(connection));
```

File: ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpServer.java (hunk 8 above)
Comment (jbrisbin) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpServer.java:24]: I wasn't thinking one would have to subscribe to start a server. I was just thinking about a callback for shutdown rather than using blocking semantics (at least as part of the abstractions of the kernel...baked in, if you will).

But that does bring up the question of codifying lifecycle events in the kernel at all. IMO these things would be better off in the transport layer. I wonder if it's possible to reduce all the things about starting and stopping a server and connecting and closing a client into a common abstraction that simply exposes start and stop. If that's all that's modeled, I'm not sure it's even necessary since the kernel components could simply accept the handlers and Publishers directly. Each transport would still have to provide special methods for configuration anyway so it seems somewhat unnecessary.

File: ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpInterceptor.java (hunk 7 above)
Comment (jbrisbin) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpInterceptor.java:5]: > I did not quiet understand, can you elaborate?

Reconnection (and connection management in general) is an important use case we have that currently powers the Spring STOMP support. We actually support dynamic reconnection to a completely different address if one is down, so whatever we do with regards to lifecycle events would have to be aware that a connection is not a one-time event and that setup and teardown of connections can happen irregardless of the lifecycle of the client.

File: ripc-rxjava1/src/main/java/io/ripc/rx/protocol/tcp/RxConnection.java (hunk 9 above)
Comment (jbrisbin) [inline ripc-rxjava1/src/main/java/io/ripc/rx/protocol/tcp/RxConnection.java:39]: I think in this case it would actually be more than that since every `Publisher` needs a `Subscriber` and a `Subscription` to be useful. It's likely some objects could be reused but if we assume the worst case (which is probably the most likely) then every write would involve the allocation of at least 5 objects + whatever wrappers, delegates, etc... exist in between (which could mean 3 \* numOfTransformations). For Reactor we address this by using a `Stream` to represent input and output (something that's only created once) and we use a `Codec` facility to manage message transformation (something that's also only create once unless not practical).

I was envisioning a "write" facility in this that isn't like `write(byte[])` but like `sink(ByteProducer)`. Namely an operation you could set up once and then pipe data into. You could always use it as a one-off operation but it would be more inefficient in that case.

File: ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpInterceptor.java (hunk 7 above)
Comment (jbrisbin) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpInterceptor.java:5]: I think there's also the question of "what exactly are we intercepting" and "to what end?". Are we intercepting **only** data messages coming in and going out? Those are two different things, actually: reading requires a `subscribe` and writing an entirely separate `Publisher` with a possibly completely different type. In the TCP case they'll often be the same (ex: `ByteBuf`) but they likely won't be in request/response servers.

If we're intercepting for the purposes of providing, say authorization, then we're doing more than just message transformation because we would be possibly closing the connection (canceling the `Subscription`).

I'm wondering if "intercepting" shouldn't be done at the site of the event (`Subscription.request` and `onNext`) and as a translation/codec/whatever rather than incurring the overhead of wrapping, delegates, etc...

File: ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpServer.java (hunk 8 above)
Comment (jbrisbin) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpServer.java:24]: I'm thinking we need to be more judicious in our use of RS components and, just like the handler, look more to simple callbacks for all situations where that's possible and only use RS components for those key critical path elements that need backpressure support. Everything else could simply leverage simply callbacks and functional transformations (e.g. `Consumer` and `Function`).

File: ripc-rxjava1/src/main/java/io/ripc/rx/protocol/tcp/RxConnection.java (hunk 9 above)
Comment (jbrisbin) [inline ripc-rxjava1/src/main/java/io/ripc/rx/protocol/tcp/RxConnection.java:39]: I'm actually starting to think that "write" is completely misleading. What we're really needing to expose is `sink(Processor<W,W>)`. `Processor.onNext(W)` would be called to do the actual writing.

Comment (jbrisbin) [inline ripc-rxjava1/src/main/java/io/ripc/rx/protocol/tcp/RxConnection.java:39]: Now that I think about (and play with) it, a `Connection` is simply a `Processor<R,W>`.

File: ripc-core/src/main/java/io/ripc/internal/Publishers.java (hunk 5 above)
Comment (NiteshKant) [inline ripc-core/src/main/java/io/ripc/internal/Publishers.java:19]: Sure, just to be clear, the implementation is not feature complete. It was a strawman to start discussion in light of all layers being implemented.

> > The Specifications class was intended to be the helper that aggregated these checks.

Not sure if having an independent Specifications class helps in this case. How does it makes sure that everyone abides by the spec?

File: build.gradle (hunk 4 above)
Comment (NiteshKant) [inline build.gradle]: Aah this was a left-over from me having the example (as a test) inside ripc-protocol-tcp demonstrating the usage. This is a test dep though. I will remove it, it isn't used.

File: build.gradle (hunk 3 above)
Comment (NiteshKant) [inline build.gradle:119]: The reason to remove "composition" was to not indicate that it is just for functional composition. In the other issue, we are discussing to rename the layer to "adapter" (or adaptor), so when we do that we can make the name change part of the project.

File: ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpInterceptor.java (hunk 2 above)
Comment (NiteshKant) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpInterceptor.java:3]: I agree that 4 generic types are verbose but I don't see how moving two to the method level will make it any different. The method definition then becomes horrific.

Finagle makes it simpler for interceptors that do not change the types, by having a `SimpleFilter` which has the same in & out types.

File: ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpHandler.java (hunk 1 above)
Comment (NiteshKant) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpHandler.java:5]: The reason I added Tcp prefix was that `Handler` by itself is confusing when someone starts creating two servers with different transports in the same code. In that scenario they have to use the fully qualified name for one of the handlers. Even `ConnectionHandler` is ambiguous when we have a `WebSocket`/`HTTP/2` connection handler.

Comment (jbrisbin) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpHandler.java:5]: > The reason I added Tcp prefix was that Handler by itself is confusing when someone starts creating two servers with different transports in the same code.

Good point.

Maybe this should be applied across-the-board so that `TcpConnection` doesn't interfere with `WebSocketConnection` et al?

File: ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpInterceptor.java (hunk 7 above)
Comment (jbrisbin) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpInterceptor.java:5]: I sketched out providing for interceptors on the connection by separating the reading and writing. This allows for different interceptors (and even multiple interceptors) on the read and write which I think will be necessary since reading and writing on a connection is often unbalanced (e.g. more emphasis is often placed on reading data and figuring out what to do than on what's written out to the client, since it can often be nothing more than an ACK).

https://gist.github.com/jbrisbin/686e565660b27b9b0f91

File: ripc-rxjava1/src/main/java/io/ripc/rx/protocol/tcp/RxConnection.java (hunk 9 above)
Comment (NiteshKant) [inline ripc-rxjava1/src/main/java/io/ripc/rx/protocol/tcp/RxConnection.java:39]: > For Reactor we address this by using a Stream to represent input and output (something that's only created once) and we use a Codec facility to manage message transformation (something that's also only create once unless not practical).

Here the object allocation is because of the transformation b/w `Publisher` and `Observable` (and other associated objects) Even with a `Stream` abstraction, how will it solve the issue? We would still have to convert from `Stream` to `Observable`. 

> I was envisioning a "write" facility in this that isn't like write(byte[]) but like sink(ByteProducer). Namely an operation you could set up once and then pipe data into.

How is `sink(ByteProducer)` different than `sink(Publisher)`?

> Now that I think about (and play with) it, a Connection is simply a Processor<R,W>

Not really because `write(Publisher)` returns the result of the write where as `onNext(Producer)` does not.

File: ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpServer.java (hunk 8 above)
Comment (NiteshKant) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpServer.java:24]: Ok. Based on this, can you suggest what changes would you make to the current `TcpServer`?

File: ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpInterceptor.java (hunk 7 above)
Comment (NiteshKant) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpInterceptor.java:5]: > Reconnection (and connection management in general) is an important use case we have that currently powers the Spring STOMP support.

Actually, it is a very mainstream usecase in almost all applications. The way we do it in Netflix is by introducing a loadbalancer that does the determination of how to retry/load balance. Since, I have not even introduced the client APIs yet, I would like to keep this discussion for the client APIs. The API here is purely for the server.

> > I think there's also the question of "what exactly are we intercepting" and "to what end?"

In this particular case (TCP server interceptors) we are intercepting "connection processing" that will include but not limited to:
- Modifying input/output.
- Introspecting input/output from the point of view of insights.
- Short-circuiting connection processing i.e. the connection does not even reach the actual handler.
- Metrics gathering of connection processing times.
- Logging

> > I'm wondering if "intercepting" shouldn't be done at the site of the event (Subscription.request and onNext) and as a translation/codec/whatever rather than incurring the overhead of wrapping, delegates, etc...

I think you are looking at it mostly from the point of view reading/writing data, in which case it may be ok to intercept read/write separately. However, that approach does not work well when we start talking about short-circuiting. The current approach of treating server as a function (issue #15) covers all the usecases. I would like to see if you have a usecase which can not be covered by that design.

> > This allows for different interceptors (and even multiple interceptors) on the read and write

The current design does not restrict the server to a single interceptor. I have described above the issue with designing interceptors as being applied on read and write.

File: ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpHandler.java (hunk 1 above)
Comment (NiteshKant) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpHandler.java:5]: > Maybe this should be applied across-the-board so that TcpConnection doesn't interfere with WebSocketConnection et al?

Indeed, it will be.

**Comment by jbrisbin** [comment]
FWIW We talked about this for 3 hours yesterday and have several concrete thoughts about some of these issues. I'm putting together some code to illustrate our thoughts and I'll try and create a summary of our discussion to point out the things we think are important. I'm just trying to figure out how the concepts and suggestions encoded in this PR converge with some of the other things we talked about and how we could bring them closer together.

TL;DR we don't think this PR is all that far away from where we probably need to be to do the kinds of things we need to do in the varied use cases we have (many of which vary quite radically from one another). That said, we do find a few things problematic and I'll try and explain why and demonstrate with code in a separate PR, which we can hopefully use as discussion points to iterate again on something that meets both needs.

**Comment by NiteshKant** [comment]
> we don't think this PR is all that far away from where we probably need to be to do the kinds of things we need to do in the varied use cases we have

Great! Eager to hear your opinions.

> That said, we do find a few things problematic and I'll try and explain why and demonstrate with code in a separate PR

Would it help if this is merged or your PR would be orthogonal to these changes?

**Comment by jbrisbin** [comment]
I think these changes will be too different to do much in the way of a merge. What I hope will happen is that we discuss the pros and cons of both PRs on their own merits and then prepare **another** PR that iterates on those ideas. That PR would then be merged to provide the next round, etc...

**Comment by NiteshKant** [comment]
@jbrisbin sounds great!

**Comment by NiteshKant** [comment]
I have removed interception and flush semantics from this PR.

File: ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpConnection.java [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+
+import org.reactivestreams.Publisher;
+
+/**
+ * An abstraction for a TCP connection.
+ *
+ * @param <R> The type of objects read from this connection.
+ * @param <W> The type of objects written to this connection.
+ */
+public interface TcpConnection<R, W> extends Publisher<R> {
```
Comment (smaldini) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpConnection.java:11]: Why having a separate tcp and ripc-core ? I would suggest just removing all Tcp\* prefixes and merge core and tcp protocol. Looking at the artifacts there is nothing specific yet to Tcp. Same for Server and Client, does it have to be in separate module ? Even in a case of File streaming I could see the same abstraction working as well.

File: ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpConnection.java [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+
+    /**
+     * Writes the passed stream of {@code data} and returns the result as a {@link Publisher}. All items emitted by
+     * this stream are flushed on completion of the stream.
+     *
+     * @param data Data stream to write.
+     *
+     * @return Result of write.
+     */
+	Publisher<Void> write(Publisher<W> data);
```
Comment (smaldini) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpConnection.java:21]: I'd prefer writeWith/writeFrom to signal the intent clearly that we don't write a publisher but from/with a publisher (writeFrom). I think writeWith removes the confusion around single writer vs multiplexing too.

File: build.gradle [hunk 12]
```diff
@@ -98,17 +113,32 @@ project('ripc-transport-netty4') {
...
+        compile 'io.reactivex:rxjava-reactive-streams:0.5.0'
+	}
+}
 
-		// RxJava 1.0
-		compile "io.reactivex:rxjava:$rxjava1Version"
+project('ripc-rxjava1-examples') {
+	description = 'Reactive IPC Composition Layer examples'
+
+    compileTestJava {
```
Comment (smaldini) [inline build.gradle]: should add compileJava compat for 1.7.

File: ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpConnection.java (hunk 10 above)
Comment (NiteshKant) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpConnection.java:11]: > Why having a separate tcp and ripc-core

Have a look at issue #18 where we are suggesting a different project structure.
ripc-core would have protocol neutral classes and different protocols will be different projects.

> Looking at the artifacts there is nothing specific yet to Tcp

The implementations are TCP specific. Connection concept is protocol specific (connection based protocols). UDP would not have a connection as an example.

> Same for Server and Client, does it have to be in separate module?

There is no client as of now. Yeah, they should be in a single module/project per protocol.

> Even in a case of File streaming I could see the same abstraction working as well.

You mean file streaming over TCP, HTTP?

File: ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpConnection.java (hunk 11 above)
Comment (NiteshKant) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpConnection.java:21]: > I'd prefer writeWith/writeFrom to signal the intent clearly that we don't write a publisher but from/with a publisher (writeFrom)

I think that would be pedantic. A `Publisher` is a source of data. I think for people dealing with asynchronous streams, it is pretty clear that in this case we are writing data from the passed source/stream and not the source itself.

>  I think writeWith removes the confusion around single writer vs multiplexing too.

How so?

File: build.gradle (hunk 12 above)
Comment (NiteshKant) [inline build.gradle]: Sure. I did not change the build file much from the version currently in the master. Let me look at it again to see what else should be modified/added.

File: ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpConnection.java (hunk 10 above)
Comment (smaldini) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpConnection.java:11]: Think about the channel from Netty then or from Java NIO, there is absolutely no need to replicate write for every sort of IO, it's not a big deal but its an API question. It would be like saying we have different type of Subscriber in Reactive Stream for Synchronous and Asynchronous ones. UDP is also handled through Channel. So maybe Channel is a better name overall if we want a common denominator.
At this module level we are API centric not Application centric anyway IMO, feel free to name it RxTcpConnection in rxjava tho if you bring specific value to this.

`You mean file streaming over TCP, HTTP?`
No, I am thinking about reading and writing a File. E.g. Chronicle or other File Memory Mapped IPC used in high throughput apps.

File: ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpConnection.java (hunk 11 above)
Comment (smaldini) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpConnection.java:21]: We are dealing with an API not an application. So yes pedantic is a qualifier it should have. Then writeWith is better than writeFrom to defuse any possibility to think it can be called once.

File: ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpConnection.java (hunk 10 above)
Comment (smaldini) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpConnection.java:11]: Although I'm not saying specific protocol don't have their modules, but we should have a common notion living in ripc core, named Connection or Channel or Pipe or Flow or IO or whatever.
If we find anything specific to TCP in addition to write and subscribe, e.g. UDP or 0MQ specific helpers they will certainly be able to extend this common artifact. It will bring much more value to the core and a bunch of candidates protocol won't even bother going that far in their first iteration. RIPC is about connecting and moving data between processes so we must make sure the core interface represents that. Server and Client can also do the same, and you already have this notion in RxNetty (as in reactor-net), another qualifier as generic IMO is Recipient and Sender. This is inherent to any communication.

Besides from the place of the components and the naming, I'm equally satisfied with the current both PR which are converging on most of the concept and I find your implementation quite smart (using the Netty user fired event to subscribe was a nice trick :+1: ). I'm in fact already working at mapping the concept in reactor-net to make the transition later easy to the whole concept. But yes as an API this is also very important to set the right expectations. I find the interception a bit more flexible in the alternative PR we issued but beyond that the concepts is really cool.

Comment (NiteshKant) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpConnection.java:11]: I see the disconnect here. I think what you are mentioning is more of a commonality in an implementation i.e. every protocol using a socket/channel to read/write data. What this PR is defining is the client/server API for different protocols to be used by the users of reactive-ipc. From an API point of view, different protocols are very different and atleast while implementing UDP, TCP, HTTP, I have not seen commonalities worth defining a common contract across these protocols. This is the reason I am removing all commonalities across protocols in RxNetty w.r.t. client/server APIs. I would recommend refraining from creating these common abstractions  unless we have multiple protocols implemented and we see value in such abstractions.

File: ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpConnection.java (hunk 11 above)
Comment (NiteshKant) [inline ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpConnection.java:21]: Let us defer this discussion for now and see if we have any other disagreement before we can merge this PR. I would like to move forward with code for now and we can always bike-shed on names later :)

File: build.gradle (hunk 12 above)
Comment (NiteshKant) [inline build.gradle]: I have made changes to the build file, have a look.

**Comment by smaldini** [comment]
Currently experimenting with this too with reactor, will update asap but looks nice, although the backpressure thing is not simple, not because of the value but because when requesting from inside netty thread it might clash with the onNext written data (some from a different thread, some from the same than netty which then applies the write directly).

I've aligned the reactor-net GA with the semantics in order to move forward easily in the future.

**Comment by NiteshKant** [comment]
@smaldini I did not implement backpressure in this PR to focus discussion on the API and we can discuss backpressure separately.
Backpressure isn't straight forward and just like you are solving it for reactor, I have solved it for RxNetty, it will be great to discuss the details of the same.

For now, I will merge this PR so that we can iterate upon specifics from here.
